core/src/main/scala/kafka/controller/PartitionStateMachineV2.scala

```diff
@@ -0,0 +1,268 @@
+      logFailedStateChange(partition, partitionState(partition), OnlinePartition, new StateChangeFailedException(failMsg))
+    }
+    val shuttingDownBrokers  = controllerContext.shuttingDownBrokerIds.toSet
+    val (partitionsWithoutLeaders, partitionsWithLeaders) = partitionLeaderElectionStrategy match {
+      case OfflinePartitionLeaderElectionStrategy =>
+        leaderForOffline(validPartitionsForElection).partition(_._2.isEmpty)
+      case ReassignPartitionLeaderElectionStrategy =>
+        leaderForReassign(validPartitionsForElection).partition(_._2.isEmpty)
+      case PreferredReplicaPartitionLeaderElectionStrategy =>
+        leaderForPreferredReplica(validPartitionsForElection).partition(_._2.isEmpty)
+      case ControlledShutdownPartitionLeaderElectionStrategy =>
+        leaderForControlledShutdown(validPartitionsForElection, shuttingDownBrokers).partition(_._2.isEmpty)
+    }
+    partitionsWithoutLeaders.foreach { case (partition, leaderIsrAndControllerEpochOpt, recipients) =>
+    }
+    val leaderIsrAndControllerEpochAndContexts = partitionsWithLeaders.map { case (partition, leaderIsrAndControllerEpochOpt, recipients) =>
+      partition -> (leaderIsrAndControllerEpochOpt.get -> (partition, leaderIsrAndControllerEpochOpt.get, recipients))
+    }
+    val setDataResponses = KafkaControllerZkUtils.setTopicPartitionStates(zookeeperClient, leaderIsrAndControllerEpochAndContexts.toMap)
+    setDataResponses.foreach { setDataResponse =>
+      val code = Code.get(setDataResponse.rc)
+      val (partition, leaderIsrAndControllerEpoch, recipients) = setDataResponse.ctx.asInstanceOf[(TopicAndPartition, LeaderIsrAndControllerEpoch, Seq[Int])]
+      if (code == Code.OK) {
+        val replicas = controllerContext.partitionReplicaAssignment(partition)
+        controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+        controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipients, partition.topic, partition.partition, leaderIsrAndControllerEpoch, replicas)
+      }
+    }
+  }
+
+  private def leaderForOffline(leaderIsrAndControllerEpochs: Seq[(TopicAndPartition, LeaderIsrAndControllerEpoch)]) = {
+    val (partitionsWithNoLiveInSyncReplicas, partitionsWithLiveInSyncReplicas) = leaderIsrAndControllerEpochs.partition { case (partition, leaderIsrAndControllerEpoch) =>
+      val liveInSyncReplicas = leaderIsrAndControllerEpoch.leaderAndIsr.isr.filter(replica => controllerContext.isReplicaOnline(replica, partition))
+      liveInSyncReplicas.isEmpty
+    }
+    val getDataResponses = KafkaControllerZkUtils.getTopicConfigs(zookeeperClient, partitionsWithNoLiveInSyncReplicas.map { case (partition, leaderIsrAndControllerEpoch) =>
+      partition.topic -> (partition, leaderIsrAndControllerEpoch)
+    }.toMap)
+    val config = controller.config.originals()
+    val partitionsWithUncleanLeaderElectionState = getDataResponses.map { getDataResponse =>
```

Comment:
partitionsWithUncleanLeaderElectionState includes partitions that don't need unclean leader election. So, the name seems a bit miss-leading.